Replace pkg/rand by standard library math/rand/v2

tklauser opened the pull request:
Since Go 1.22 the Go standard library provides the `math/rand/v2` package[^1]. Its global generator accessed by top-level functions is unconditionally seeded using a true random value. Moreover, the new package provides a way to construct PRNG sources which can be accessed concurrently.

This allows to drop `SafeRand` and use the global source again where applicable. `SafeRand` was originally introduced in commit fac5dde ("rand: add and use concurrency-safe PRNG source") to fix #10988 by providing a concurrency-safe PRNG source other than the global `math/rand` source which could be used at the time because it can't be interfered with from unrelated packages by means of calling rand.Seed, see #10575.

In some cases the global `math/rand/v2` source cannot be used because unit tests rely on seeding the PRNG source to a fixed value for deterministic test runs. In these cases a `math/rand/v2` source using fixed values is used, like with `SafeRand` before.

[^1]: https://go.dev/doc/go1.22#math_rand_v2

See commits for details.

Supersedes #29731

tklauser committed:
rand: replace with math/rand/v2

Since Go 1.22 the Go standard library provides the math/rand/v2
package[^1]. Its global generator accessed by top-level functions is
unconditionally seeded using a true random value. Moreover, the new
package provides a way to construct PRNG sources which can be accessed
concurrently.

This allows to drop SafeRand and use the global source again where
applicable. SafeRand was originally introduced in commit fac5dde ("rand:
add and use concurrency-safe PRNG source") to fix #10988 by providing a
concurrency-safe PRNG source other than the global math/rand source
which could be used at the time because it can't be interfered with from
unrelated packages by means of calling rand.Seed, see #10575.

In some cases the global math/rand/v2 source cannot be used because unit
tests rely on seeding the PRNG source to a fixed value for deterministic
test runs. In these cases a math/rand/v2 source using fixed values is
used, like with SafeRand before.

[^1]: https://go.dev/doc/go1.22#math_rand_v2

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
rand: remove package

The previous commit changed all users of pkg/rand to use math/rand/v2
from the Go standard library. As a result this package is no longer
needed, should not be used going forward and can thus be removed.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
make: remove script checking for pkg/rand usage

pkg/rand was removed in the previous commit and we'll rely on
math/rand/v2 from the Go standard library going forward. The script
checking for pkg/rand usage is thus no longer needed.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

sayboras commented on daemon/cmd/status.go:
```diff
@@ -6,6 +6,7 @@ package cmd
 import (
 	"context"
 	"fmt"
+	rand "math/rand/v2"
```
nit: the alias here is redundant

tklauser replied:
Yes, this is intentional.  I'm not sure we have any coding style guidelines regarding named imports. Personally I prefer to add an explicit alias in case the base name of the import (i.e. `v2` in this case) doesn't match the package name. But that might be redundant for versioned packages :thinking:

bimmlerd replied:
I also prefer `rand` to `v2`, but I think Go indeed does this by default, so the alias isn't necessary: https://go.dev/play/p/Esxcqx8uaGR

tklauser replied:
Thanks for checking. I'll drop the alias since this seems to be the common pattern e.g. in the Go standard libary itself as well.

sayboras reviewed (approved):
Thanks and lgtm :heavy_check_mark:

giorio94 commented on pkg/proxy/proxy.go:
```diff
@@ -23,15 +24,14 @@ import (
 	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "proxy")
 
-	portRandomizer = rand.NewSafeRand(time.Now().UnixNano())
+	portRandomizer = rand.New(rand.NewPCG(uint64(time.Now().UnixNano()), 0))
```
This generator doesn't seem to be ever re-seeded. The global one could be probably used instead.

tklauser replied:
Good catch, will use the global functions in this package instead.

giorio94 commented on pkg/ipam/service/allocator/bitmap.go:
```diff
@@ -51,7 +51,7 @@ type bitAllocator interface {
 	a := AllocationBitmap{
 		strategy: randomScanStrategy{
-			rand: rand.NewSafeRand(time.Now().UnixNano()),
+			rand: rand.New(rand.NewPCG(uint64(time.Now().UnixNano()), 0)),
```
Could the global random generator be used here as well?

tklauser replied:
Yeah, let's do that.

giorio94 commented on pkg/lock/stoppable_waitgroup_test.go:
```diff
@@ -151,16 +151,12 @@ func TestWaitChannel(t *testing.T) {
 		for {
 			select {
-			case in <- rand.Intn(1 - 0):
+			case in <- rand.IntN(1 - 0):
```
Unrelated to this PR, but this doesn't seem that useful as it will always return 0.

giorio94 reviewed (commented):
Changes overall look good to me, thanks!

Just a couple of comments inline. I wonder if it wouldn't be better to convert the few final `math/rand` imports over to v2, and then add a linter to forbid using `math/rand` altogether, to prevent possible regressions.

tklauser committed:
Use math/rand/v2 instead of math/rand

Replace use of the standard library math/rand package by math/rand/v2
which provides top-level functions backed by a global generator
unconditionally seeded using a true random value. This avoids having to
create generators manually.

Suggested-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
multicast: don't use deprecated math/rand.Read

Use crypto/rand.Read instead as suggested in
https://pkg.go.dev/math/rand#Read

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
golangci-lint: ensure math/rand/v2 is used instead of math/rand

Use the depguard linter to disallow using the standard library math/rand
package and use v2 of the package instead.

Suggested-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

giorio94 reviewed (approved):
Thanks!

pippolo84 reviewed (approved):
FQDN changes LGTM, thanks! :100:

christarazi reviewed (approved):
LGTM for ci-structure
